# 出力、変数のフォーマット

In [4]:
a = 123
print(str(a).zfill(5))
print(f'{a:05}')
print(f'{str(a):05}')
b = -123
print(str(b).zfill(5))
print(f'{b:05}')
print(f'{str(b):05}')
c = 1.23
print(str(c).zfill(5))
print(f'{c:05}')
print(f'{str(c):05}')


00123
00123
12300
-0123
-0123
-1230
01.23
01.23
1.230


### numpyやfloatの数値末尾の処理
np.format_float_positional

In [2]:
import numpy as np

SAVE_FILE = 'format_test.txt'

step = 0.1
start = 5.0
size = 51
data = []
for i in range(size):
    data.append((step * i) + start)

u_data = []
f_data = []
t_0_data = []
t_p_data = []
t_m_data = []
t_m8_data = []
print(data)

for i,value in enumerate(data):
    data[i] = np.format_float_positional(value, precision=14)
    u_data.append(np.format_float_positional(value, precision=14, unique=False))
    f_data.append(np.format_float_positional(value, precision=14, fractional=False))
    # Controls post-processing trimming of trailing digits
    t_0_data.append(np.format_float_positional(value, precision=14, trim='0'))
    t_p_data.append(np.format_float_positional(value, precision=14, trim='.'))
    t_m_data.append(np.format_float_positional(value, precision=14, trim='-'))
    t_m8_data.append(np.format_float_positional(value, precision=8, trim='-'))
print(data)
print(u_data)
print(f_data)
print(t_0_data)
print(t_p_data)
print(t_m_data)
print(t_m8_data)

print(*data)

with open(SAVE_FILE, 'w') as fs:
    fs.write(str(data))
    fs.write('\n')
    print(f_data, file=fs)
    print(t_0_data, file=fs)
    print(t_p_data, file=fs)
    print(t_m_data, file=fs)
    print(*data, file=fs)
    



[5.0, 5.1, 5.2, 5.3, 5.4, 5.5, 5.6, 5.7, 5.8, 5.9, 6.0, 6.1, 6.2, 6.3, 6.4, 6.5, 6.6, 6.7, 6.8, 6.9, 7.0, 7.1, 7.2, 7.300000000000001, 7.4, 7.5, 7.6, 7.7, 7.800000000000001, 7.9, 8.0, 8.1, 8.2, 8.3, 8.4, 8.5, 8.6, 8.7, 8.8, 8.9, 9.0, 9.100000000000001, 9.2, 9.3, 9.4, 9.5, 9.600000000000001, 9.7, 9.8, 9.9, 10.0]
['5.', '5.1', '5.2', '5.3', '5.4', '5.5', '5.6', '5.7', '5.8', '5.9', '6.', '6.1', '6.2', '6.3', '6.4', '6.5', '6.6', '6.7', '6.8', '6.9', '7.', '7.1', '7.2', '7.30000000000000', '7.4', '7.5', '7.6', '7.7', '7.80000000000000', '7.9', '8.', '8.1', '8.2', '8.3', '8.4', '8.5', '8.6', '8.7', '8.8', '8.9', '9.', '9.10000000000000', '9.2', '9.3', '9.4', '9.5', '9.60000000000000', '9.7', '9.8', '9.9', '10.']
['5.00000000000000', '5.10000000000000', '5.20000000000000', '5.30000000000000', '5.40000000000000', '5.50000000000000', '5.60000000000000', '5.70000000000000', '5.80000000000000', '5.90000000000000', '6.00000000000000', '6.10000000000000', '6.20000000000000', '6.30000000000000', '

In [9]:
import numpy as np
a = 2545.0
print(type(a))
print(np.format_float_positional(a, precision=14, trim='-'))
data = [2545]
b = np.array(data, dtype='float32')
print(b)
print(np.format_float_positional(b, precision=14, trim='-'))
c = np.array(data, dtype='float64')
print(c)
print(np.format_float_positional(c, precision=14, trim='-'))

np.set_printoptions(suppress=True, precision=6, floatmode='unique')
np.savetxt('f32.txt', b)


<class 'float'>
2545
[2545.]
2545
[2545.]
2545


### ndarrayの整数末尾の"."が取れないとき  
再現できないのはnumpyのバージョンが原因かもしれない


In [30]:
import numpy as np
note = 'test\n'
DIGIT_PRECISION = 8
a = [[0.0500000000001819, 2525.0]]
nda = np.array(a, dtype='float64')
axis = []
buf_axis = nda
print(f'{buf_axis}:{type(buf_axis)}')
for axis_values in buf_axis:
    step = axis_values[0]
    print(f'{step}:{type(step)}')
    start = axis_values[-1]
    print(f'{start}:{type(start)}')
    division = 401
    end = step * (division - 1) + start
    print(f'{end}:{type(end)}')
    axis.append([start, end, step, division])
print(f'{axis}:{type(axis)}')
dimension_start = axis[0][0]
note += f'Dimension start={np.format_float_positional(dimension_start, precision=DIGIT_PRECISION, trim="-")}\n'
test_end = axis[0][1]
print(type(test_end))
# note += f'Dimension {axis_count} end={np.format_float_positional(axis[i][1], precision=self.DIGIT_PRECISION, trim="-")}\n'
# XAFSプログラムのpxt2txt.pyではここでend=2525.になるが再現できない
note += f'Dimension  end={np.format_float_positional(test_end, precision=DIGIT_PRECISION, trim="-")}\n'

dimension_step = axis[0][2]
note += f'Dimension step={np.format_float_positional(dimension_step, precision=DIGIT_PRECISION, trim="-")}\n'
dimension_size = axis[0][3]
note += f'Dimension size={dimension_size}\n'
print(note)
axis[0][0].astype(np.float32)

[[5.000e-02 2.525e+03]]:<class 'numpy.ndarray'>
0.0500000000001819:<class 'numpy.float64'>
2525.0:<class 'numpy.float64'>
2545.0000000000728:<class 'numpy.float64'>
[[2525.0, 2545.0000000000728, 0.0500000000001819, 401]]:<class 'list'>
<class 'numpy.float64'>
test
Dimension start=2525
Dimension  end=2545
Dimension step=0.05
Dimension size=401



2525.0

In [28]:
import numpy as np
fi32 = np.finfo(np.float32)
fi64 = np.finfo(np.float64)
print(fi32)
print(fi64)
print(np.__version__)

Machine parameters for float32
---------------------------------------------------------------
precision =   6   resolution = 1.0000000e-06
machep =    -23   eps =        1.1920929e-07
negep =     -24   epsneg =     5.9604645e-08
minexp =   -126   tiny =       1.1754944e-38
maxexp =    128   max =        3.4028235e+38
nexp =        8   min =        -max
smallest_normal = 1.1754944e-38   smallest_subnormal = 1.4012985e-45
---------------------------------------------------------------

Machine parameters for float64
---------------------------------------------------------------
precision =  15   resolution = 1.0000000000000001e-15
machep =    -52   eps =        2.2204460492503131e-16
negep =     -53   epsneg =     1.1102230246251565e-16
minexp =  -1022   tiny =       2.2250738585072014e-308
maxexp =   1024   max =        1.7976931348623157e+308
nexp =       11   min =        -max
smallest_normal = 2.2250738585072014e-308   smallest_subnormal = 4.9406564584124654e-324
------------------

pxt2txtのnumpyバージョンは 1.19.5  
バージョンの違いか
'''
---------------------------------------------------------------
precision =  15   resolution = 1.0000000000000001e-15
machep =    -52   eps =        2.2204460492503131e-16
negep =     -53   epsneg =     1.1102230246251565e-16
minexp =  -1022   tiny =       2.2250738585072014e-308
maxexp =   1024   max =        1.7976931348623157e+308
nexp =       11   min =        -max
'''

### str -> int等の変換

In [2]:
str_int = '0001'
int_data = int(str_int)
print(int_data)

1


### ndaarryの表示
- ndarryとlistではprint表示に違いがある
- ndarryは小数点以下が0の時.だけで終わる
- float32からfloat64への変換は切りの悪い数値になる  
floatによるキャストが良い

In [3]:
import numpy as np
a = [790.15, 790.16, 790]
b = np.array(a, dtype=np.float32)
for value in b:
     print(np.format_float_positional(value, precision=14, trim='-'))
c = []

for value in a:
    float_value = float(value)
    c.append(float_value)
    print(np.format_float_positional(float_value, precision=14, trim='-'))
d = b.astype(np.float64)
e = np.array(c)
f = np.array(a, dtype=np.float64)

print(a)
print(b)
print(c)
print(d)
print(e)
print(f)
np.set_printoptions(suppress=False, precision=8, floatmode='fixed')
print('set options')
print(d)
print(e)
print(f)


790.15
790.16
790
790.15
790.16
790
[790.15, 790.16, 790]
[790.15002441 790.15997314 790.00000000]
[790.15, 790.16, 790.0]
[790.15002441 790.15997314 790.00000000]
[790.15000000 790.16000000 790.00000000]
[790.15000000 790.16000000 790.00000000]
set options
[790.15002441 790.15997314 790.00000000]
[790.15000000 790.16000000 790.00000000]
[790.15000000 790.16000000 790.00000000]


### "C:\Users\fujikawa\Documents\works\ARIM\XPS\IGORテストファイル\軟X線光電子分光装置\SES_angle.txt"のデータの精度

- 少なくとも小数点第8位以下は精度不足で補完された値
- 装置出力のもともとの精度はfloat32

In [27]:
import numpy as np

# numpyの精度を指定
np.set_printoptions(suppress=False, precision=12, floatmode='fixed')

# XPS装置からの出力結果
original_values = [4.80055332183838E+0000,
                   7.67665290832520E+0000,  6.77240550518036E-0001]
print(f'装置排出値\n{original_values}')
float64_original_values = np.array(original_values, dtype=np.float64)
print(f'{float64_original_values}: {float64_original_values.dtype}')

# 小数点第8位以下を消去
values = [4.8005533,
          7.6766529,  0.67724055]
print(f'ダミー\n{values}')
float32_values = np.array(values, dtype=np.float32)
print(f'{float32_values}: {float32_values.dtype}')




装置排出値
[4.80055332183838, 7.6766529083252, 0.677240550518036]
[4.800553321838 7.676652908325 0.677240550518]: float64
ダミー
[4.8005533, 7.6766529, 0.67724055]
[4.800553321838 7.676652908325 0.677240550518]: float32
